# Copyright (C) 2016-2020 Maciej Delmanowski <drybjed@gmail.com>
# Copyright (C) 2016-2020 DebOps <https://debops.org/>
# SPDX-License-Identifier: GPL-3.0-only

# During 'vagrant package' operation, 'virt-sysprep' command removes existing
# SSH host keys. However, at the next boot of the packaged box, they are not
# regenerated automatically, and this breaks Vagrant provisioning as well as
# the 'vagrant ssh' command. Another problem is re-creation of the existing
# '/etc/machine-id' file if it has been zeroed out, which breaks DHCP and VM
# networking during operation of multiple Vagrant machines at once.
#
# This patch is hopefully a temporary solution to these problems until a better
# one is found. It will disable the removal of existing SSH host keys and VM
# customization by the 'virt-sysprep' command. This is not advisable in
# a production environment, however can be sufficient for a temporary, test
# environment.
#
# See also: https://github.com/vagrant-libvirt/vagrant-libvirt/issues/759
# See also: https://github.com/vagrant-libvirt/vagrant-libvirt/issues/851#issuecomment-520799720
#
# Work directory: /usr/share/rubygems-integration/all/gems/vagrant-libvirt-*

--- lib/vagrant-libvirt/action/package_domain.rb
+++ lib/vagrant-libvirt/action/package_domain.rb
@@ -38,7 +38,7 @@
           `qemu-img rebase -p -b "" #{@tmp_img}`
           # remove hw association with interface
           # working for centos with lvs default disks
-          `virt-sysprep --no-logfile --operations defaults,-ssh-userdir -a #{@tmp_img}`
+          `virt-sysprep --no-logfile --operations defaults,-ssh-userdir,-ssh-hostkeys,-customize -a #{@tmp_img}`
           Dir.chdir(@tmp_dir)
           img_size = `qemu-img info #{@tmp_img} | grep 'virtual size' | awk '{print $3;}' | tr -d 'G'`.chomp
           File.write(@tmp_dir + '/metadata.json', metadata_content(img_size))
